### ISSUE TITLE
Recheck executes test way more times than expected

### ISSUE THREAD

**Issue by AlexeyRaga** [dharmaturtle/fsharp-hedgehog-xunit#7]
This executes the test once:

```csharp
    [Fact]
    public void ShouldMap_UserCompetition_Equivalent2()
    {
        var prop =
            from source in ForAll(GenX.autoWith<CompetitionDetailsCache>(CompetitionsAutoGenConfig.Config()))
            select source.ToUserCompetition(Mapper)
                         .Should()
                         .BeEquivalentTo(
                             source.ToUserCompetition());

        prop.ReCheck("50_17003736010433171956_2975322824293632487_000000000000000000101010111110111111011111110111111110111111111101111111111101111111111110111111111111101111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110");
    }
```

And this executes the test `128667` times, always with the same values:
```csharp
    [Property, Recheck("50_17003736010433171956_2975322824293632487_000000000000000000101010111110111111011111110111111110111111111101111111111101111111111110111111111111101111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110111111111111111110")]
    public void ShouldMap_UserCompetition_Equivalent(CompetitionDetailsCache source)
    {
        source.ToUserCompetition(Mapper)
              .Should()
              .BeEquivalentTo(source.ToUserCompetition());
    }
```

Changing a number of tests in `Property` attribute (I tried `[Property(1)]`) doesn't seem to have any affect on this behaviour.

Expected: test case is executed once in both cases :)

**Comment by dharmaturtle** [comment]
I believe this is an [upstream issue](https://github.com/hedgehogqa/fsharp-hedgehog/issues/420).

**Comment by dharmaturtle** [comment]
I no longer believe this is an upstream issue. The problem is with [this code](https://github.com/dharmaturtle/fsharp-hedgehog-xunit/blob/393a064a0504940e614dfa9f52b870fa87832172/src/Hedgehog.Xunit/InternalLogic.fs#L152-L169) - I need to return unit from inside the Property to satisfy the CE. Currently not sure how to do that. My best attempt

```fsharp
      |> toProperty // line 162
      |> Property.bind (fun () ->
        property {
          return () // compilation error
        }
      )
```

has the following compilation error:

> This expression was expected to have type    'bool'    but here has type    'unit'

Using `return true` instead of `return ()` compiles, but still runs multiple times with Rechecking, so I'm not even sure if I'm on the right track.

@TysonMN

**Comment by TysonMN** [comment]
Maybe lines 169-172 should be

```fsharp
Property.check <| property {
  let! args = gens
  inkove args
  return ()
}
```

**Comment by TysonMN** [comment]
But that can't be right because it is ignoring logic that rechecks or not.  I also don't understand `toProperty` right now, which seems like an important function.  The last line of `inkove` is to call `toProperty`.

**Comment by dharmaturtle** [comment]
`testMethod` is the body of a method that has been decorated with the `[<Property>]` attribute. On [161](https://github.com/dharmaturtle/fsharp-hedgehog-xunit/blob/393a064a0504940e614dfa9f52b870fa87832172/src/Hedgehog.Xunit/InternalLogic.fs#L161) I `Invoke` that method. `Invoke` returns whatever that method returns. The return may be a bool, in which case I throw it into `Property.ofBool`. Or it may be an Async/Task, in which case I await it. It goes recursive because there may be a second Task in the Task. Or it may be a `Result<_,_>` in which case it asserts that the Result is in the OK state.

If it is none of these, it's a `Property.success ()`. This case is relevant to this thread since Alexy's test returns unit/void.

The [invoke on 152](https://github.com/dharmaturtle/fsharp-hedgehog-xunit/blob/393a064a0504940e614dfa9f52b870fa87832172/src/Hedgehog.Xunit/InternalLogic.fs#L152) only serves as an arg to [`Property.forAll`](https://github.com/dharmaturtle/fsharp-hedgehog-xunit/blob/393a064a0504940e614dfa9f52b870fa87832172/src/Hedgehog.Xunit/InternalLogic.fs#L169). If we move away from `Property.forAll`, I'd still like some way to match over the return type of the test method.

**Comment by TysonMN** [comment]
So I think the code I suggested should be "pushed down". Do you know where?

I _think_ we don't want to call `Property.forall`. I think you want to call `Property.check` or `Property.recheck` (it some variant) instead.

**Comment by dharmaturtle** [comment]
[`BindReturn`](https://github.com/hedgehogqa/fsharp-hedgehog/blob/b48f662ba42541dd51b0a116cca2489339d14d6a/src/Hedgehog/Property.fs#L331) works - [I empirically verified that it only runs once for a sample test](https://user-images.githubusercontent.com/1278803/181925981-6f8318d3-2266-43cc-a331-79693241bcdb.png).

Two possible issues, though neither are insurmountable:
* It shortcuts `toProperty`. This may be resolved by changing `toProperty` to return unit. (Also means a name update.) To avoid a breaking change, `Property.ofBool` could do `Xunit.Assert.True`. Feels a little hacky, but it works...
* `BindReturn` uses methods that are `internal` to Hedgehog core.

**Comment by TysonMN** [comment]
I expect the fix is a change to this project.  I would be very surprised if the proper fix was a change to Hedgehog.

**Comment by TysonMN** [comment]
You treat `bool` special because Hedgehog treats `bool` special?

I am thinking of removing the special treatment of `bool`. https://github.com/hedgehogqa/fsharp-hedgehog/issues/419#issuecomment-1200430718

Or, I could throw a custom exception if the value is `false`.

**Comment by dharmaturtle** [comment]
> You treat bool special because Hedgehog treats bool special?

Yes.

Though I wouldn't necessarily drop support in this library if Hedgehog drops special treatment of bool. My philosophy with this library is to try to make the experience of devs easier - it is, after all, "Hedgehog with _convenience_ attributes for xUnit.net". Correctness is your domain :) I already do special casing for tests that return async/task/result. Why not bool?

> Or, I could throw a custom exception if the value is `false`.

Hah - I know we talked about using `Xunit.Assert.True` to replace [Property.ofBool](https://github.com/dharmaturtle/fsharp-hedgehog-xunit/commit/bd81548c5b56d34661020db7936a3ded3a92ba1e#diff-e465cd73e9c7ed48576c99c20d6f3fae146c512a0bb9e07cf6ce357374bdf31dL93). Independent of your comment, while writing [the test](https://github.com/dharmaturtle/fsharp-hedgehog-xunit/commit/baf1c5831ee6892ac112ea5d193d4a01780b684f), I decided against that and wrote a [new custom exception](https://github.com/dharmaturtle/fsharp-hedgehog-xunit/commit/bd81548c5b56d34661020db7936a3ded3a92ba1e#diff-d698d8f1828ddc03cd54f4d1940431f44c513400873f7380a0db598ecdf7af7d). I feel like it makes it more apparent in the failing stacktrace what's going on.
